From 918c12dc823f258ca2fde6a7c36ab60f2a7ced99 Mon Sep 17 00:00:00 2001
From: Macross Chen <macross.chen@mediatek.com>
Date: Thu, 25 May 2023 11:39:57 +0800
Subject: [PATCH] Do not show payload info if show_payload flag off

If the ctrl type is V4L2_CTRL_TYPE_STRING,
v4l2-ctl will set flags with V4L2_CTRL_FLAG_HAS_PAYLOAD.

If the ctrl flags are with V4L2_CTRL_FLAG_WRITE_ONLY,
v4l2-ctl will not create any payload buffer for query results.

In the V4L2_CTRL_FLAG_WRITE_ONLY + V4L2_CTRL_TYPE_STRING case,
print_value() be called with show_payload=false, but v4l2-ctl
still try to access the payload(string) buffer.

Just don't print value while show_payload == false.

Signed-off-by: Macross Chen <macross.chen@mediatek.com>
---
 utils/v4l2-ctl/v4l2-ctl-common.cpp | 3 ++-
 1 file changed, 2 insertions(+), 1 deletion(-)

diff --git a/utils/v4l2-ctl/v4l2-ctl-common.cpp b/utils/v4l2-ctl/v4l2-ctl-common.cpp
index 49c0f39..76f56c0 100644
--- a/utils/v4l2-ctl/v4l2-ctl-common.cpp
+++ b/utils/v4l2-ctl/v4l2-ctl-common.cpp
@@ -495,6 +495,8 @@ static void print_value(int fd, const v4l2_query_ext_ctrl &qc, const v4l2_ext_co
 	}
 
 	if (qc.flags & V4L2_CTRL_FLAG_HAS_PAYLOAD) {
+		if (!show_payload)
+			return;
 		switch (qc.type) {
 		case V4L2_CTRL_TYPE_U8:
 			printf("%u", *ctrl.p_u8);
@@ -1232,7 +1234,6 @@ void common_get(cv4l_fd &_fd)
 						print_value(fd, qc, ctrl, true, true);
 						return;
 					}
-
 					printf("%s: ", name.c_str());
 					print_value(fd, qc, ctrl, true, false);
 					printf("\n");
-- 
2.18.0

